README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

# Custom Camera Plugin

[![pub package](https://img.shields.io/pub/v/custom_cam.svg)](https://pub.dev/packages/custom_cam)

A Flutter plugin for iOS and Android to allow an easy custom UI camera set-up with photo and video capabilites that stores files locally internal to the app and returns its path and media type.

|                | Android | iOS      |
|----------------|---------|----------|
| **Support**    | SDK 21+ | iOS 10+* |

## Features

* Camera permission exception handling.
* Display live camera preview.
* Device orientation responsive UI.
* Record video.
* Take photos.
* Zoom with gestures.
* Persistant storage trough app files.

## Getting started

First, add `custom_cam` as a [dependency in your pubspec.yaml file](https://flutter.dev/using-packages/).

### iOS

\* The camera plugin compiles for any version of iOS, but its functionality
requires iOS 10 or higher. If compiling for iOS 9, make sure to programmatically
check the version of iOS running on the device before using any camera plugin features.
The [device_info_plus](https://pub.dev/packages/device_info_plus) plugin, for example, can be used to check the iOS version.

Add two rows to the `ios/Runner/Info.plist`:

* one with the key `Privacy - Camera Usage Description` and a usage description.
* and one with the key `Privacy - Microphone Usage Description` and a usage description.

If editing `Info.plist` as text, add:

```xml
<key>NSCameraUsageDescription</key>
<string>your usage description here</string>
<key>NSMicrophoneUsageDescription</key>
<string>your usage description here</string>
```

### Android

Change the minimum Android sdk version to 21 (or higher) in your `android/app/build.gradle` file.

```groovy
minSdkVersion 21
```

## Usage

Here is a small example of a flutter app that opens the camera and waits for the camera result. MultimediaItem is a package class with two parameters: one for the storage path and another to differentiate between video and photo.

```dart
final MultimediaItem? result = await Navigator.of(context).push(
        MaterialPageRoute(builder: (_) => const CustomCamera(
            primaryColor: Color(0xff000fff), 
            secondaryColor: Color(0xfffff0fff), 
            backgroundColor: Color(0xff000faa)
        ))
);
```

## Additional information

For a more elaborate usage example see [here](https://github.com/marcelosolorzano/custom_cam/tree/main/example).


CHANGELOG.md
# 1.0.4

* Fix custom icons not being displayed

# 1.0.3

* Remove hardcoded text font.

# 1.0.2

* Fix alert dialog content padding.

# 1.0.1

* Provide a photo-only mode.
* Replace generic font with custom.
* Implement screen util tool to adapt screen and font size.
* Minor UI improvements.

## 1.0.0

* Initial release.


lib/custom_cam.dart
library custom_cam;

import 'package:camera/camera.dart';
import 'package:custom_cam/src/custom_icons_icons.dart';
import 'package:custom_cam/src/camera_alert.dart';
import 'package:custom_cam/src/custom_theme.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';
import '/src/multimedia_item.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'src/camera_preview.dart' as camera_preview;

export 'src/multimedia_item.dart' show MultimediaItem;

class CustomCamera extends StatefulWidget {

  final Color primaryColor;
  final Color secondaryColor;
  final Color backgroundColor;
  final bool isRecordingEnabled;

  const CustomCamera({Key? key, required this.primaryColor, required this.secondaryColor, required this.backgroundColor, this.isRecordingEnabled = false}) : super(key: key);

  @override
  State<CustomCamera> createState() => _CustomCameraState();
}

class _CustomCameraState extends State<CustomCamera> with WidgetsBindingObserver {

  List<CameraDescription> cameras = <CameraDescription>[];

  CameraController? controller;

  late bool _isVideoMode = false;
  late bool _isVideoRecording = false;

  double _minAvailableZoom = 1.0;
  double _maxAvailableZoom = 1.0;
  double _currentScale = 1.0;
  double _baseScale = 1.0;

  // Counting pointers (number of user fingers on screen)
  int _pointers = 0;

  Future<void> initCamera() async {
    exitCallback() => { Navigator.of(context).pop() };

    // initialize cameras.
    try {
      cameras = await availableCameras();
    }
    on CameraException catch (e) {
      CameraAlert exitAlert = CameraAlert(title: 'Ocurrió un error', description: 'No fue posible inicializar la cámara', positiveInput: 'Aceptar', negativeInput: '', positiveCallback: exitCallback);
      showDialog(context: context, builder: (_) { return exitAlert; });
      debugPrint("$e occurred while initializing the camera");
    }


    final CameraController? oldController = controller;
    if (oldController != null) {
      // `controller` needs to be set to null before getting disposed,
      // to avoid a race condition when we use the controller that is being
      // disposed. This happens when camera permission dialog shows up,
      // which triggers `didChangeAppLifecycleState`, which disposes and
      // re-creates the controller.
      controller = null;
      await oldController.dispose();
    }

    final CameraController cameraController = CameraController(cameras[0], ResolutionPreset.high, enableAudio: widget.isRecordingEnabled);

    controller = cameraController;

    // Update UI if controller updated
    cameraController.addListener(() {
      if (mounted) setState(() {});
    });

    try {
      await cameraController.initialize();
      await Future.wait(<Future<Object?>>[
        cameraController
            .getMaxZoomLevel()
            .then((double value) => _maxAvailableZoom = value),
        cameraController
            .getMinZoomLevel()
            .then((double value) => _minAvailableZoom = value),
      ]);
    } on CameraException catch (e) {
      String errorDescription;
      switch (e.code) {
        case 'CameraAccessDenied':
          errorDescription = 'Los permisos de cámara son necesarios para usar esta funcionalidad';
          break;
        case 'CameraAccessDeniedWithoutPrompt':
          // iOS only
          errorDescription = 'Por favor conceda los permisos de cámara desde la configuración de la aplicación';
          break;
        case 'CameraAccessRestricted':
          // iOS only
          errorDescription = 'El acceso a la cámara se encuentra restringido';
          break;
        case 'AudioAccessDenied':
          errorDescription = 'Los permisos de grabación de audio son necesarios para usar esta funcionalidad';
          break;
        case 'AudioAccessDeniedWithoutPrompt':
          // iOS only
          errorDescription = 'Por favor conceda los permisos de grabación de audio desde la configuración de la aplicación';
          break;
        case 'AudioAccessRestricted':
          // iOS only
          errorDescription = 'El acceso a la grabación de audio se encuentra restringido';
          break;
        default:
          errorDescription = 'No fue posible inicializar la cámara';
          break;
      }
      CameraAlert exitAlert = CameraAlert(title: 'Ocurrió un error', description: errorDescription, positiveInput: 'Aceptar', negativeInput: '', positiveCallback: exitCallback);
      if (mounted) showDialog(context: context, builder: (_) { return exitAlert; });
    }

    if (mounted) {
      setState(() {});
    }
  }

  @override
  void initState() {
    super.initState();
    CustomTheme.primaryColor = widget.primaryColor;
    CustomTheme.secondaryColor = widget.secondaryColor;
    CustomTheme.backgroundColor = widget.backgroundColor;
    // initialize the rear camera
    initCamera();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    final CameraController? cameraController = controller;

    // App state changed before we got the chance to initialize.
    if (cameraController == null || !cameraController.value.isInitialized) {
      return;
    }

    if (state == AppLifecycleState.inactive) {
      cameraController.dispose();
    } else if (state == AppLifecycleState.resumed) {
      initCamera();
    }
  }

  @override
  void dispose() {
    // Dispose of the controller when the widget is disposed.
    controller?.dispose();
    super.dispose();
  }

  Future<void> takePicture() async {
    final CameraController? cameraController = controller;

    if (!cameraController!.value.isInitialized) {return;}
    if (cameraController.value.isTakingPicture) {return;}
    try {
      lockDeviceOrientation();
      await cameraController.setFlashMode(FlashMode.off);
      XFile picture = await cameraController.takePicture();
      goToPreview(picture.path, false);
    }
    on CameraException catch (e) {
      debugPrint('Error occurred while taking picture: $e');
      return;
    }
  }

  void lockDeviceOrientation() {
    List<DeviceOrientation> deviceOrientation = MediaQuery.of(context).orientation == Orientation.portrait ? [DeviceOrientation.portraitUp, DeviceOrientation.portraitDown] : [DeviceOrientation.landscapeLeft, DeviceOrientation.landscapeRight];
    SystemChrome.setPreferredOrientations(deviceOrientation);
  }

  Future<void> startVideoRecording() async {
    final CameraController? cameraController = controller;

    if (!cameraController!.value.isInitialized) {return;}
    if (cameraController.value.isRecordingVideo) {return;}
    try {
      lockDeviceOrientation();
      await cameraController.startVideoRecording();
    }
    on CameraException catch (e) {
      debugPrint('Error occurred while starting to record video: $e');
      return;
    }
  }

  Future<void> stopVideoRecording() async {
    final CameraController? cameraController = controller;

    if (!cameraController!.value.isInitialized) {return;}
    if (!cameraController.value.isRecordingVideo) {return;}
    try {
      XFile video = await cameraController.stopVideoRecording();
      goToPreview(video.path, true);
    }
    on CameraException catch (e) {
      debugPrint('Error occurred while stopping video record: $e');
      return;
    }
  }

  void goToPreview(String url, bool isVideo) async {
    final result = await Navigator.of(context).push(MaterialPageRoute(builder: (_) => camera_preview.CameraPreview(multimediaItem: MultimediaItem(url, isVideo))));
    if (result is MultimediaItem) {
      if (mounted) Navigator.of(context).pop(result);
    }
  }

  Widget _getVideoControls() {
    return widget.isRecordingEnabled ?
      Expanded(
          child: TextButton(
            style: CustomTheme.circularButtonStyle,
            onPressed: () {
              setState(() => _isVideoMode = !_isVideoMode);
            },
            child: Icon(
                !_isVideoMode
                    ? Icons.videocam
                    : Icons.camera_alt,
                color: Colors.white,
                size: 30.w),
          )
      ) : const Spacer();
  }

  @override
  Widget build(BuildContext context) {
    ScreenUtil.init(context, designSize: const Size(375, 812));
    return Scaffold(
        body: SafeArea(
            child:
            Stack(
                children: [
                  _cameraPreviewWidget(),
                  OrientationBuilder(builder: (context, orientation) {
                    return Align(
                      alignment: orientation == Orientation.portrait ?  Alignment.topRight : Alignment.topLeft,
                      child: IconButton(onPressed: () {
                        exitCallback() => { Navigator.of(context).pop() };
                        CameraAlert exitAlert = CameraAlert(title: 'Salir de fotografías', description: 'Al salir perderá la información ingresada y no podrá recuperarla. ¿Desea continuar?', positiveInput: 'Salir', negativeInput: 'Volver', positiveCallback: exitCallback);
                        showDialog(context: context, builder: (_) { return exitAlert; });
                      }, icon: Icon(CustomIcons.close, size: 23.w, color: CustomTheme.secondaryColor)),
                    );
                  }),
                  OrientationBuilder(
                      builder: (context, orientation) {
                        return Align(
                            alignment: orientation == Orientation.portrait ? Alignment.bottomCenter : Alignment.centerRight,
                            child: Container(
                              height: orientation == Orientation.portrait ? 189.h : null,
                              width: orientation == Orientation.portrait ? null : 189.w,
                              decoration: BoxDecoration(color: CustomTheme.backgroundColor.withOpacity(0.8)),
                              child:
                              Flex(direction: orientation == Orientation.portrait ? Axis.horizontal : Axis.vertical,
                                  crossAxisAlignment: CrossAxisAlignment.center,
                                  children: [
                                    _getVideoControls(),
                                    Expanded(
                                        child: TextButton(
                                          onPressed: () {
                                            if (!_isVideoMode) {
                                              takePicture();
                                            }
                                            else {
                                              if (_isVideoRecording) {
                                                stopVideoRecording();
                                              }
                                              else {
                                                startVideoRecording();
                                              }

                                              setState(() => _isVideoRecording = !_isVideoRecording);
                                            }
                                          },
                                          style: CustomTheme.circularButtonStyle,
                                          child: SizedBox(
                                            width: 82.w,
                                            height: 82.w,
                                            child: Icon(
                                                !_isVideoMode
                                                    ? (widget.isRecordingEnabled ? Icons.camera : CustomIcons.camera)
                                                    : !_isVideoRecording ? Icons.fiber_manual_record : Icons.stop,
                                                color: Colors.white,
                                                size: widget.isRecordingEnabled ? 50.w : 45.w),
                                          ),
                                        )),
                                    const Spacer(),
                                  ]),
                            )
                        );
                      }
                  )
                ])
        )
    );
  }

  /// Display the preview from the camera (or a message if the preview is not available).
  Widget _cameraPreviewWidget() {
    final CameraController? cameraController = controller;

    if (cameraController == null || !cameraController.value.isInitialized) {
      return const Center(child: CircularProgressIndicator());
    } else {
      return Listener(
        onPointerDown: (_) => _pointers++,
        onPointerUp: (_) => _pointers--,
        child: CameraPreview(
          controller!,
          child: LayoutBuilder(
              builder: (BuildContext context, BoxConstraints constraints) {
                return GestureDetector(
                    behavior: HitTestBehavior.opaque,
                    onScaleStart: _handleScaleStart,
                    onScaleUpdate: _handleScaleUpdate
                );
              }),
        ),
      );
    }
  }

  void _handleScaleStart(ScaleStartDetails details) {
    _baseScale = _currentScale;
  }

  Future<void> _handleScaleUpdate(ScaleUpdateDetails details) async {
    // When there are not exactly two fingers on screen don't scale
    if (controller == null || _pointers != 2) {
      return;
    }

    _currentScale = (_baseScale * details.scale)
        .clamp(_minAvailableZoom, _maxAvailableZoom);

    await controller!.setZoomLevel(_currentScale);
  }
}



lib/src/camera_alert.dart
import 'package:custom_cam/src/custom_icons_icons.dart';
import 'package:custom_cam/src/custom_theme.dart';
import 'package:flutter/material.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';

class CameraAlert extends StatelessWidget {

  final String title;
  final String description;
  final String positiveInput;
  final String negativeInput;
  final VoidCallback positiveCallback;

  const CameraAlert({Key? key, required this.title, required this.description, required this.positiveInput, required this.negativeInput, required this.positiveCallback}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      title: Padding(
        padding: EdgeInsets.only(top: 16.h),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Icon(CustomIcons.warning, size: 55.w, color: CustomTheme.secondaryColor),
            Padding(
              padding: EdgeInsets.only(top: 18.h),
              child: SizedBox(width: 250.w, child: Text(title, textAlign: TextAlign.center, style: TextStyle(color: const Color(0xFF333333), fontSize: 18.sp, fontWeight: FontWeight.w700))),
            )
          ],
        ),
      ),
      content: SizedBox(width: 250, child: Text(description, textAlign: TextAlign.center, style: TextStyle(color: const Color(0xFF2C2C2C), fontSize: 14.sp, fontWeight: FontWeight.w400))),
      actionsAlignment: MainAxisAlignment.center,
      actions: [
        Padding(
          padding: EdgeInsets.only(bottom: 30.h),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              SizedBox(
                width: 250.w,
                child: TextButton(
                    onPressed: () {
                      Navigator.of(context).pop();
                      positiveCallback();
                    },
                    style: CustomTheme.textButtonStyle,
                    child: Padding(
                      padding: const EdgeInsets.all(6.0),
                      child: Text(positiveInput),
                    )
                ),
              ),
              if (negativeInput.isNotEmpty)
                TextButton(
                    onPressed: () async {
                      Navigator.of(context).pop();
                    },
                    style: TextButton.styleFrom(
                        foregroundColor: CustomTheme.primaryColor
                    ),
                    child: Text(negativeInput)
                )
            ],
          ),
        )
      ],
      elevation: 0,
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.all(Radius.circular(20.r))
      ),
    );
  }
}


lib/src/custom_theme.dart
import 'package:flutter/material.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';

class CustomTheme {
  static Color primaryColor = const Color(0xFF004691);
  static Color secondaryColor = const Color(0xFFF18700);
  static Color backgroundColor = const Color(0xFFFFFFFF);

  static ButtonStyle textButtonStyle = TextButton.styleFrom(
      foregroundColor: Colors.white,
      disabledForegroundColor: primaryColor,
      backgroundColor: primaryColor,
      shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(6.0)
      ),
      textStyle: TextStyle(
        fontSize: 16.sp,
        fontWeight: FontWeight.w600
      )
  );

  static ButtonStyle circularButtonStyle = TextButton.styleFrom(
      foregroundColor: Colors.white,
      disabledForegroundColor: primaryColor,
      backgroundColor: primaryColor,
      shape: const CircleBorder(),
      textStyle: TextStyle(
        fontSize: 16.sp,
        fontWeight: FontWeight.w600
      )
  );
}

lib/src/camera_preview.dart
import 'dart:io';

import 'package:custom_cam/src/custom_icons_icons.dart';
import 'package:custom_cam/src/multimedia_item.dart';
import 'package:custom_cam/src/custom_theme.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_screenutil/flutter_screenutil.dart';
import 'package:path_provider/path_provider.dart';
import 'package:video_player/video_player.dart';

import 'camera_alert.dart';

class CameraPreview extends StatefulWidget {

  final MultimediaItem multimediaItem;

  const CameraPreview({Key? key, required this.multimediaItem}) : super(key: key);

  @override
  State<CameraPreview> createState() => _CameraPreviewState();
}

class _CameraPreviewState extends State<CameraPreview> {
  VideoPlayerController? _videoController;

  @override
  void initState() {
    super.initState();

    // Create and store the VideoPlayerController. The VideoPlayerController
    // offers several different constructors to play videos from assets, files,
    // or the internet.
    if (widget.multimediaItem.isVideo) {
      startVideoPlayer();
    }
  }

  Future startVideoPlayer() async {
    _videoController = VideoPlayerController.file(File(widget.multimediaItem.path));
    await _videoController!.initialize().then((_) {
      // Ensure the first frame is shown after the video is initialized,
      // even before the play button has been pressed.
      setState(() {});
    });
    await _videoController!.setLooping(true);
    await _videoController!.play();
  }

  Future saveToDocuments() async {
    int currentUnix = DateTime.now().millisecondsSinceEpoch;

    final directory = await getApplicationDocumentsDirectory();
    String fileFormat = widget.multimediaItem.path.split('.').last;

    await File(widget.multimediaItem.path).copy('${directory.path}/$currentUnix.$fileFormat',);
  }

  @override
  void dispose() {
    SystemChrome.setPreferredOrientations([
      DeviceOrientation.landscapeRight,
      DeviceOrientation.landscapeLeft,
      DeviceOrientation.portraitUp,
      DeviceOrientation.portraitDown,
    ]);
    // Ensure disposing of the VideoPlayerController to free up resources.
    _videoController?.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: SafeArea(
          child: Stack(
            children: [
              !widget.multimediaItem.isVideo ? Image.file(File(widget.multimediaItem.path))
                  : SizedBox(
                width: double.infinity, height: double.infinity,
                child: AspectRatio(
                  aspectRatio: _videoController!.value.aspectRatio,
                  child: GestureDetector(
                    child: VideoPlayer(_videoController!),
                    onTap: () {
                      // Wrap the play or pause in a call to `setState`. This ensures the
                      // correct icon is shown.
                      setState(() {
                        // If the video is playing, pause it.
                        if (_videoController!.value.isPlaying) {
                          _videoController!.pause();
                        } else {
                          // If the video is paused, play it.
                          _videoController!.play();
                        }
                      });
                    },
                  ),
                ),
              ),
              Align(
                alignment: Alignment.topRight,
                child: IconButton(onPressed: () {
                  exitCallback() => { Navigator.of(context).pop() };
                  CameraAlert exitAlert = CameraAlert(title: 'Salir de fotografías', description: 'Al salir perderá la información ingresada y no podrá recuperarla. ¿Desea continuar?', positiveInput: 'Salir', negativeInput: 'Volver', positiveCallback: exitCallback);
                  showDialog(context: context, builder: (_) { return exitAlert; });
                }, icon: Icon(CustomIcons.close, size: 23.w), color: CustomTheme.secondaryColor),
              ),
              OrientationBuilder(
                  builder: (context, orientation) {
                    return Align(
                        alignment: orientation == Orientation.portrait ? Alignment.bottomCenter : Alignment.centerRight,
                        child: Container(
                          height: orientation == Orientation.portrait ? 189.h : double.infinity,
                          width: orientation == Orientation.portrait ? double.infinity : 189.w,
                          decoration: BoxDecoration(color: CustomTheme.backgroundColor.withOpacity(0.8)),
                          child:
                          Column(
                              mainAxisAlignment: MainAxisAlignment.center,
                              children: [
                                SizedBox(
                                  width: 250.w,
                                  child: TextButton(
                                      onPressed: () {
                                        saveToDocuments();
                                        Navigator.of(context).pop(widget.multimediaItem);
                                      },
                                      style: CustomTheme.textButtonStyle,
                                      child: const Padding(
                                        padding: EdgeInsets.all(6.0),
                                        child: Text('Guardar'),
                                      )
                                  ),
                                ),
                                TextButton(
                                    onPressed: () async {
                                      Navigator.of(context).pop();
                                    },
                                    style: TextButton.styleFrom(
                                        foregroundColor: CustomTheme.primaryColor
                                    ),
                                    child: Text('Volver a tomar', style: TextStyle(fontSize: 14.sp, fontWeight: FontWeight.w700))
                                ),
                              ]),
                        )
                    );
                  }
              )
            ],
          ),
        )
    );
  }
}


lib/src/custom_icons_icons.dart
/// Flutter icons CustomIcons
/// Copyright (C) 2023 by original authors @ fluttericon.com, fontello.com
/// This font was generated by FlutterIcon.com, which is derived from Fontello.
///
/// To use this font, place it in your fonts/ directory and include the
/// following in your pubspec.yaml
///
/// flutter:
///   fonts:
///    - family:  CustomIcons
///      fonts:
///       - asset: fonts/CustomIcons.ttf
///
/// 
///
import 'package:flutter/widgets.dart';

class CustomIcons {
  CustomIcons._();

  static const _kFontFam = 'CustomIcons';
  static const String _kFontPkg = 'custom_cam';

  static const IconData camera = IconData(0xe800, fontFamily: _kFontFam, fontPackage: _kFontPkg);
  static const IconData close = IconData(0xe801, fontFamily: _kFontFam, fontPackage: _kFontPkg);
  static const IconData warning = IconData(0xe803, fontFamily: _kFontFam, fontPackage: _kFontPkg);
}


example/lib/main.dart
import 'package:custom_cam/custom_cam.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'CustomCam Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'CustomCam Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String _counter = '';

  Future<void> _openCamera() async {
    final MultimediaItem? result = await Navigator.of(context).push(
        MaterialPageRoute(builder: (_) => const CustomCamera(
            primaryColor: Color(0xff000fff),
            secondaryColor: Color(0xfffff0fff),
            backgroundColor: Color(0xff000faa)
        ))
    );
    if (result != null) {
      setState(() {
        _counter = result.path;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'A new file was saved to:',
            ),
            Text(
              _counter,
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _openCamera,
        tooltip: 'Camera',
        child: const Icon(Icons.camera_alt),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
